### ISSUE TITLE
Fully sync cache with repo

### ISSUE THREAD

**Issue by richardolsson** [zetkin/lyra#25]
![image](https://github.com/zetkin/lyra/assets/550212/f0229326-f653-4999-a4fb-c5ccf5e963eb)

**Comment by WULCAN** [comment]
On e5d32971b88e6576ce5f87db8df10ca9d0a8b564, translations are loaded from the local git clone on first use, and then never reloaded until the Lyra server process restarts.

The local git clone is updated from the remote before and after creating a pull-request so the translators' changes will be applied on a fresh copy, regardless of what the translator saw when she made the translations. The translator will never see this until the Lyra server process restarts, as the changes from the remote are never loaded into Lyra.

Try for example:
1. Start Lyra targeting a project named p with a supported language code la and a base branch b
2. Open /projects/p/la
3. Make changes translation or the messages in the remote base_branch
4. Reload /projects/p/la: you will not see the new changes
5. Save some changes to the translations on /projects/p/la and create a pull request.
6. Reload /projects/p/la: you will still not see the changes from the remote.

This is a problem because if we add new messages to a project, Lyra will have to be restarted to pick up those changes. When we restart Lyra, we risk losing translators' work that was made after the last pull request was created.

It is also a problem when we change parameters for a message. Translators will keep translating with the old parameters until Lyra is restarted.

It is also a problem when we change the default message of a message. Translators will keep  translating the old, stale default message until Lyra is restarted.

Richard's diagram in this issue is a proposed solution that can resolve these problems. The diagram is already added to Lyra's repository, but the implementation does not match it yet.

**Comment by WULCAN** [comment]
The translation store is used to support a small set of HTTP operations:
* Load all projects' messages and languages to display progress bars on home page
* Load one project's messages and languages to display progress bars on project page
* Load one project's messages and languages to display translations in translation UI
* Update a single translation in a server action
* Create Pull request in a server action

We image a single interface, to gate getting messages and translations (and likely to update them and to create pull requests) This could be Cache or Store.

It will need a merge function that merges the previous state from a new state read from GitHub via git and a file system.

We add an internal method Revalidate to Cache/Store, that is used by its public methods like getTranslations and getMessages. This method pulls from the remote, re-reads translations from fs, and then merges and updates storage.

Then we add logic to skip reading from fs if pull did not find a new revision.

Then we add logic to skip pull if we pulled recently.

Neither updateTranslations nor sendPullRequest should do any of this, but still need to access translations. They need special access via Cache, or even better, Cache needs those two features. Cache will need to use RepoGit, but RepoGit should not need to use Cache.

In summary:

1. Move most of updateTranslation to Cache
2. Move most of sendPullRequest and saveLanguageFiles from RepoGit to Cache
2. Implement Merge(old: StoreData, new StoreData): StoreData
3. Implement and add Revalidate to Cache/Store API methods like getMessages
4. Skip read and merge if already based on HEAD
5. Skip Pull if too recent

We will want to move messages into this Cache/Store too, as it is also read form remote, via fs, and also benefits from being refreshed... and for more reasons. But we do not need this before the steps above.

1. Add messages caching into Cache/Store
3. Use Cache/Store to get messages instead of directly
